### PR TITLE
Add `any` support for autograph with dataset

### PR DESCRIPTION
**Pull request by yongtang**
This PR tries to address the issue where `any` in autograph
does not work with tf.data:
```
import tensorflow as tf
def g():
  dataset = tf.data.Dataset.from_tensor_slices([False, True, False])
  return any(dataset)
print("G: ", g())
@tf.function
def f():
  dataset = tf.data.Dataset.from_tensor_slices([False, True, False])
  return any(dataset)
print("F: ", f())
..........
    OperatorNotAllowedInGraphError: using a `tf.Tensor` as a Python `bool` is not allowed: AutoGraph did not convert this function. Try decorating it directly with @tf.function.
```

This PR translate `any(dataset)` to
```
dataset.filter(lambda x: x).take(1).reduce
```
which works in tf.data's pipeline.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/python/autograph/operators/py_builtins.py [hunk 1]
```diff
@@ -384,8 +384,29 @@ def _py_filter(function, iterable):
...
+  return _py_any(iterable)
+
+
+# any() operation is essentially a "if first True element exist".
+# For that it could be translated to `filter(True)` to filter out
+# only `True` element, and then `take(1)`. This works in tf.data
+# as tf.data's filter+take is done in pipeline so it will stop
+# as soon as `take(1)` returns.
+def _tf_dataset_any(iterable):
+  return iterable.filter(
```
Comment (mdanatg) [inline tensorflow/python/autograph/operators/py_builtins.py]: The chaining causes a wrapping that's a bit awkward. Consider breaking it on multiple lines:

```
ds = iterable.filter(lambda x: x)
ds = ds.take(1)
... etc.
```

File: tensorflow/python/autograph/operators/py_builtins.py [hunk 2]
```diff
@@ -384,8 +384,29 @@ def _py_filter(function, iterable):
...
+      lambda x: x).take(1).reduce(
+          constant_op.constant(False, dtype=dtypes.bool),
+          lambda x, y: math_ops.logical_or(x, y))
```
Comment (mdanatg) [inline tensorflow/python/autograph/operators/py_builtins.py]: Strictly speaking, I don't think we need the `logical_or`, right?

* if the result is `False`, then the dataset is empty and returns the default `False`
* if the result is `True`, then the dataset has exactly one element and that is always `True`, thanks to the preceding `filter`

So this should work (with this briefly explained in a comment):

`reduce(constant_op.constant(False, dtype=dtypes.bool), lambda _, y: y)`

WDYT?

**Review by mdanatg — DISMISSED** [review]
Very nice! I would welcome a PR that adds the `all` counterpart.

File: tensorflow/python/autograph/operators/py_builtins.py [hunk 3]
```diff
@@ -384,8 +384,29 @@ def _py_filter(function, iterable):
...
+    return _tf_dataset_any(iterable)
+  return _py_any(iterable)
+
+
+# any() operation is essentially a "if first True element exist".
+# For that it could be translated to `filter(True)` to filter out
+# only `True` element, and then `take(1)`. This works in tf.data
+# as tf.data's filter+take is done in pipeline so it will stop
+# as soon as `take(1)` returns.
+def _tf_dataset_any(iterable):
```
Comment (mdanatg) [inline tensorflow/python/autograph/operators/py_builtins.py]: A great addition would be a verification that looks at `iterable.element_spec` and raises an error if the dataset contains anything other than a single `bool` scalar tensor. The error message can be something like this:  `in graph mode, the "any" builtin only supports datasets that return bool scalars; got: {}`.format(iterable.element_spec)

File: tensorflow/python/autograph/operators/py_builtins.py (hunk 1 above)
Comment (yongtang) [reply tensorflow/python/autograph/operators/py_builtins.py]: Thanks @mdanatg, the chaining has been removed.

File: tensorflow/python/autograph/operators/py_builtins.py (hunk 2 above)
Comment (yongtang) [reply tensorflow/python/autograph/operators/py_builtins.py]: Thanks @mdanatg, that makes sense. PR has been updated.

File: tensorflow/python/autograph/operators/py_builtins.py (hunk 3 above)
Comment (yongtang) [reply tensorflow/python/autograph/operators/py_builtins.py]: Thanks. The PR has been updated to check and makes sure element_spec only consists of tf.bool.

**Review by mdanatg — APPROVED** [review]
Thank you for the improvements!

**Comment by yongtang** [comment]
@mdanatg Thanks for the review! I will create a follow up PR once this PR is merged.

**Comment by yongtang** [comment]
@mdanatg  (update)  I will create a follow up PR  for `all` as well, once this PR is merged.

File: tensorflow/python/autograph/operators/py_builtins_test.py [hunk 4]
```diff
@@ -293,6 +293,34 @@ def test_filter_dataset(self):
...
+    self.assertEqual(
+        py_builtins.any_([False, True, False]), True)
+    self.assertEqual(
+        py_builtins.any_([False, False, False]), False)
+
+  def test_any_dataset(self):
+    dataset_1 = dataset_ops.DatasetV2.from_tensor_slices([False, True, False])
+    dataset_2 = dataset_ops.DatasetV2.from_tensor_slices([False, False, False])
+    self.assertEqual(
+        py_builtins.any_(dataset_1), True)
```
Comment (mdanatg) [inline tensorflow/python/autograph/operators/py_builtins_test.py]: Looking at the test results, it seems that this needs a `self.evaluate`.

**Review by mdanatg — CHANGES_REQUESTED** [review]

File: tensorflow/python/autograph/operators/py_builtins_test.py (hunk 4 above)
Comment (yongtang) [reply tensorflow/python/autograph/operators/py_builtins_test.py]: Thanks @mdanatg, PR Updated.

**Comment by yongtang** [comment]
@mdanatg The PR has been updated. Please let me know if there are any other issues.

**Review by mdanatg — APPROVED** [review]

**Pull request merged**